libraries/AC_AttitudeControl/AC_PosControl.cpp

```diff
@@ -546,125 +586,67 @@ void AC_PosControl::calc_leash_length_z()
 // vel_up_max, vel_down_max should have already been set before calling this method
 void AC_PosControl::run_z_controller()
 {
-    float curr_alt = _inav.get_altitude();
-
-    // clear position limit flags
-    _limit.pos_up = false;
-    _limit.pos_down = false;
-
-    // calculate altitude error
-    _pos_error.z = _pos_target.z - curr_alt;
-
-    // do not let target altitude get too far from current altitude
-    if (_pos_error.z > _leash_up_z) {
-        _pos_target.z = curr_alt + _leash_up_z;
-        _pos_error.z = _leash_up_z;
-        _limit.pos_up = true;
-    }
-    if (_pos_error.z < -_leash_down_z) {
-        _pos_target.z = curr_alt - _leash_down_z;
-        _pos_error.z = -_leash_down_z;
-        _limit.pos_down = true;
-    }
-
-    // calculate _vel_target.z using from _pos_error.z using sqrt controller
-    _vel_target.z = sqrt_controller(_pos_error.z, _p_pos_z.kP(), _accel_z_cms, _dt);
-
-    // check speed limits
-    // To-Do: check these speed limits here or in the pos->rate controller
-    _limit.vel_up = false;
-    _limit.vel_down = false;
-    if (_vel_target.z < _speed_down_cms) {
-        _vel_target.z = _speed_down_cms;
-        _limit.vel_down = true;
-    }
-    if (_vel_target.z > _speed_up_cms) {
-        _vel_target.z = _speed_up_cms;
-        _limit.vel_up = true;
-    }
+    // Position Controller
 
+    float curr_alt = _inav.get_altitude();
+    // define maximum position error and maximum first and second differential limits
+    _p_pos_z.set_limits_error(fabsf(_leash_down_z), _leash_up_z, fabsf(_speed_down_cms), _speed_up_cms);
+    // calculate the target velocity correction
+    _vel_target.z = _p_pos_z.update_all(_pos_target.z, curr_alt, _limit.pos_down, _limit.pos_up);
     // add feed forward component
-    if (_flags.use_desvel_ff_z) {
-        _vel_target.z += _vel_desired.z;
-    }
+    _vel_target.z += constrain_float(_vel_desired.z, -fabsf(_speed_down_cms), _speed_up_cms);
```

Comment:
Agreaa with PB, all lots of `fabsf(_speed_down_cms)` going on, I'm not sure, maybe `max(_speed_down_cms,0)` would be better, or  `max(_speed_down_cms, some reasonable min value)`

Reply:
This is used because the sign has been reversed on occasion resulting in nasty results. I too would prefer that it was guaranteed to be expressed as a positive number. Human nature and all....